### ISSUE TITLE
Building DRY helpers for mutations

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/action#1216]
now that we're getting into some fun trickle-down mutations, I think it's really important to keep things logical. Let's consider the example of:
- leaving a github repo, 
- receiving a webhook from github saying you got kicked out of an organization, 
- removing a repo.

All 3 of these mutations are actually pretty similar.
- Leaving a github repo, if there are no other folks on that repo, will trigger removing a repo
- getting kicked out of a github org is a plural of leaving a github repo-- you leave many repos
- removing a repo does the following:
  - sets isActive to false
  - archives all the projects associated with that repo

There is definitely some major overlap in code, and as the business logic changes, we wanna keep the code DRY to minimize hard to track down bugs. As a second priority, we want to maximize performance, returning a  payload in short amount of time.

So, a few questions:
- Should we reuse subscription channels & payloads? If we do, it means, we'll have to send more frames to the clients, but they payloads will be simpler. See #1142. I still think 1 payload is simpler, but we can compromise by making the larger payload contain the smaller paylods, maybe in array format. that should keep sanity on the backend & frontend in terms of updating the schema & updating the store, respectively.
- how to handle a webhook? a webhook is a mutation that you didn't know you wanted, so it behaves similarly to a subscription. however, a webhook could represent an array of changes. getting kicked out of a github org is a perfect example of that. do we create a a payload exclusively for the webhook or do we send multiple leaveIntegration frames? Per #1142, the answer should be the former where the payload is simply a GraphQLList(LeavingIntegrationPayload). This makes me wonder, what if the `integrationLeft` subscription always returns GraphQLList(LeavingIntegrationPayload)? Most of the time, the length would be 1, but in doing so, we could reduce the # of channels, subscription handlers, etc, list of subscriptions attached to each QueryRenderer, etc. This is a slippery slope though, because we still want the mutationPayload == subscriptionPayload. it makes no sense for a "removeTeamMember" mutation to return an array of removed team members, although that's not a huge concern. However, that equality limits us to still only returning a single large payload instead of many smaller ones, and knowing that we could deliver 2 instead of n+1 is tempting enough to make me want to break that equality between subs & mutations, but i know that's bad, so i'm thinking i'll just make a GitHubOrgMemberRemovedWebhookPayload. The only question that remains is how to make sure that all components listening to RepoRemoved also listen to LeaveIntegration(GITHUB) and GitHubOrgMemberRemoved. 

1 solution is to add a `parentSubs` option to `ensureSubscription` & in each subscription, we list out the parents (or a thunk to compute the parents). that way RepoRemoved is always subscribed to whenever LeaveIntegration(GITHUB) is. With this inheritance pattern, it could get sticky though, so it'd be nice to automate that somehow based on "oh, you want a sub that has a nested Payload which belongs to another sub, lemme get that for you, too". but then again, a given payload doesn't have to infer a given subscription. plus, in order to do that, we'd have to do some static analysis, and writing babel code sucks.

tl;dr
- if a webhook introduces a new payload, we could either send multiple frames of the existing payload, or create a new type & use that. we opt for the latter because that affords us the option of it being more than an array
- to offset the growing number of channels, `ensureSubscription` should take a `parentSubs` thunk or array to make sure that we don't forget a new subscription on a QueryRenderer somewhere & create a ridiculously hard to find bug
- pubsub publishing should only occur once per mutation & should not occur in helpers.
